### ISSUE TITLE
CA1043: Support System.Index and System.Range

### ISSUE THREAD

**Issue by skarllot** [dotnet/roslyn-analyzers#3638]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0

#### Diagnostic ID

[CA1043](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1043)

#### Repro steps

```csharp
class Test
{
    public object this[Index index] => null;
    public object[] this[Range range] => null;
}
```

#### Expected behavior
Allow to use `System.Index` and `System.Range` as indexer argument.

#### Actual behavior
Forces to use integral type or string.

**Comment by Evangelink** [comment]
That seems like a good idea. @bartonjs Do you know if guidelines will be updated to "allow" use of `Index` and `Range` for indexers?

**Comment by bartonjs** [comment]
Range based indexers are supposed to return the same type as the method that has them, but they get pattern-matched to a Slice method, so they shouldn't really need to be defined (but if they get explicitly defined they should adhere to the guideline).

FDGv3 says

> AVOID indexers with parameter types other than System.Int32,
System.Int64, System.String, **System.Range, System.Index**, or an
enum, except on dictionary-like types.

(emphasis mine)

**Comment by Evangelink** [comment]
@bartonjs Thanks for the confirmation! I see that you mentioned `enum` but not `object`. Do you think the rule should also be updated to:
1. Remove exception for `object`
2. Add exception for `enum`
It seems weird to have a difference between the guidelines and the analyzer recommendations.

**Comment by Evangelink** [comment]
The current implementation is actually pretty different from what the doc states...

https://github.com/dotnet/roslyn-analyzers/blob/master/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseIntegralOrStringArgumentForIndexers.cs#L34-L43

**Comment by Evangelink** [comment]
@mavasani @bartonjs There were some pending questions open, let me know if you think they deserve being addressed (in which case I'll create the issue(s) and PR(s)). To sum-up the points:

1. Doc states it's ok to use the type `object` but the guidelines doesn't. Shall we update the rule or is it some extra point that might be worth having in the guideline?

2. Guidelines allows `enum` but code and doc don't. Shall we update something?

3. Guidelines + doc mention `Int32` and `Int64` but the code also allows for `Int16`, `UInt16`, `UInt32` and `UInt64`. Shall we update the doc or the code?
